### ISSUE TITLE
Simplify the React Hooks example

### ISSUE THREAD

**Issue by michaelrambeau** [davidcetinkaya/embla-carousel-react#3]
Hello again David @davidcetinkaya 

I checked the example that uses React hooks.
I'd like to discuss about 2 points to make the example simpler.

## Use ref instead of state

I noticed the example uses the state to keep a reference of the `embla` object.

Why not use the hook `useRef` to store the reference instead? (I think it's the purpose of the `useRef` hook).

It would look this:

```jsx
const EmblaCarouselComponent = () => {
  const emblaRef = useRef();
  const scrollPrev = () => emblaRef.current.scrollPrev();
  const scrollNext = () => emblaRef.current.scrollNext();

  return (
    <>
      <EmblaCarouselReact
        htmlTagName="div"
        emblaRef={embla => (emblaRef.current = embla)}
        options={{ loop: false }}
      >
        <div style={{ display: "flex" }}>
          <div style={{ flex: "0 0 100%" }}>Slide 1</div>
          <div style={{ flex: "0 0 100%" }}>Slide 2</div>
          <div style={{ flex: "0 0 100%" }}>Slide 3</div>
        </div>
      </EmblaCarouselReact>
      <button onClick={scrollPrev}>Previous</button>
      <button onClick={scrollNext}>Next</button>
    </>
  );
};
```

I removed `useCallback` since the example can work without (maybe you could provide a more sophisticated example in addition of the basic example)

I'm still not 100% satisfied by the API, the ref handling still looks a bit weird to me. Somehow the API should provide directly an `embla` object (using the render props pattern, or providing a `useEmbla` hook?) to make things even simpler... maybe? But let's keep things simple for now 😃 .

## Default props

To make the example cleaner , I think you should set default props in the carousel implementation.

Adding this to `EmblaCarouselReact` class...

```js
static defaultProps = {
  htmlTagName: "div"
};
```

...makes the example even shorter.

```jsx
const EmblaCarouselComponent = () => {
  const emblaRef = useRef();
  const scrollPrev = () => emblaRef.current.scrollPrev();
  const scrollNext = () => emblaRef.current.scrollNext();

  return (
    <>
      <EmblaCarouselReact
        emblaRef={embla => (emblaRef.current = embla)}
        options={{ loop: false }}
      >
        <div style={{ display: "flex" }}>
          <div style={{ flex: "0 0 100%" }}>Slide 1</div>
          <div style={{ flex: "0 0 100%" }}>Slide 2</div>
          <div style={{ flex: "0 0 100%" }}>Slide 3</div>
        </div>
      </EmblaCarouselReact>
      <button onClick={scrollPrev}>Previous</button>
      <button onClick={scrollNext}>Next</button>
    </>
  );
};
```

**Comment by readeral** [comment]
I appreciate this Michael, there doesn't seem to me to be any need for memoizing the embla methods - am I missing something in the inclusion of useCallback for the example as it stands? Can they not simply be arrow functions?

Regarding useRef - David's approach of putting the embla object in state makes the API closer to the like embla-carousel approach, in that there is no 'current' property needed with embla-carousel-react providing it's own ref.

I suppose it depends whether the 'useRef' hook is familiar enough to end users who have looked through the embla-carousel docs to intuitively pop in the 'current' property.

**Comment by michaelrambeau** [comment]
Hello @readeral thank you for the feedback.

I see what you mean, the ref handling with the `current` property does not feel very natural.

Thinking further, all the user needs is an instance of the `embla` object, the ref handling should be internal.

Why not such an API (for the next major release):

```jsx
import {Carousel, Slide} from 'embla-carousel-react'

<Carousel>
  {embla => (
    <>
      <Slide>Slide 1</Slide>
      <Slide>Slide 2</Slide>
      <Slide>Slide 3<Slide>
      <button onClick={() => embla.scrollPrev()}>Previous</button>
      <button onClick={() => embla.scrollNext()}>Previous</button>
    </>
  )}
</Carousel>
```

The `Carousel` object not only builds the container, it also provides the `embla` object and registers the carousel slides.

However `embla-react-carousel` would be more than just a tiny wrapper around `embla`... so it's just an idea.

**Comment by xiel** [comment]
Hey @michaelrambeau ,
just adding my two cents. The current approach using emblaRef enables a more flexible and cleaner use than a render prop imho. But in theory both apis could be easily supported side by side.

Using useRef instead of useState would actually save one initial extra render (after setEmbla is called), but that does not really matter much. useState makes usage within callbacks and useEffect simpler.

I would also inline the prev/next arrow handlers, no need to useCallback here (actually creates more overhead than savings).

**Comment by davidcetinkaya** [comment]
Hello @michaelrambeau, @readeral, @xiel,

Thank Michael you for opening this issue! I hope my answer will clear some stuff up and give you an answer of why I made the `emblaRef` design choice.

### Default props
I've opened issue #4 for the **default htmlTagName** prop suggestion. I think this is a great idea. Thanks a lot 🌟!

### Ref & State
The current implementation with the `emblaRef` prop expects a callback which returns the **Embla API** when it is **ready to use**. The reason why I use a callback is that the callback can **guarantee** the user that the vanilla JS Embla API for this carousel is actually ready to use, in contrast to just a reference (or am I wrong here?). Using **embla** as a **state** also gives the user a clear way to handle it when it's `null`. But I might be missing something? And maybe there's a better way to do it?

### Render props
I've considered render props but, like the example @michaelrambeau provided:

```javascript
import {Carousel, Slide} from 'embla-carousel-react'

<Carousel>
  {embla => (
    <>
      <Slide>Slide 1</Slide>
      <Slide>Slide 2</Slide>
      <Slide>Slide 3<Slide>
      <button onClick={() => embla.scrollPrev()}>Previous</button>
      <button onClick={() => embla.scrollNext()}>Previous</button>
    </>
  )}
</Carousel>
```

But there are some issues with this. The above code locks the **embla API** within the `<Carousel>` tag which makes assumptions about your markup regarding the slides and also the hierarchy. What if I want my **buttons** or my **dot navigation** outside the `<Carousel>` tag?

And I do actually want my buttons outside the `<Carousel>` tag. This is because the carousel viewport listens to pointer events so every time you click on a button it will also trigger a `pointerdown` event grabbing the carousel if it's already moving. This [CodeSandbox](https://codesandbox.io/s/znjzv) should demonstrate the markup freedom you get with the current approach in contrast to a render prop implementation.

### Use Callback
The reason why the example shows this:

```javascript
const scrollPrev = useCallback(() => emblaRef.current.scrollPrev(), [embla]);
const scrollNext = useCallback(() => emblaRef.current.scrollNext(), [embla]);
```

...instead of this:

```javascript
const scrollPrev = () => emblaRef.current.scrollPrev();
const scrollNext = () => emblaRef.current.scrollNext();
```

...is that as far as I understand about React Hooks (and I may be wrong), is that `scrollPrev` and `scrollNext` functions will only change if the `embla` variable changes. Without `useCallback()`, these functions would re-run on every re-render?

**Comment by gunnarx2** [comment]
### Use Callback
As @davidcetinkaya pointed out, the `useCallback()` will return a memoized version of the callback that only changes if `embla` changes. This is an optimization, but the question is if it's necessary for the example.

I agree with @xiel, for this simple example I would inline the onClick handlers. It will look more compressed and understandable.

So I would go with:
```
<button onClick={() => embla.scrollPrev()}>Previous</button>
<button onClick={() => embla.scrollNext()}>Next</button>
```

### Ref & State
`useRef()` is probably the "correct" way of doing this, `useState()` causes re-render and `useRef()` does not, but is it necessary?

I think the `useState()` is more simple/clean and you don't always have to refer to `current`.

**Comment by davidcetinkaya** [comment]
Hello all (@michaelrambeau, @xiel, @readeral, @gunnarx2),

## Conclusion
Considering the discussion in this issue:

### Use Callback
I think we can all agree on simplifying the **hooks example** by removing the `useCallbacks` and inlining the `scrollPrev` and `scrollNext` functions like so:

```jsx
<button onClick={() => embla.scrollPrev()}>Previous</button>
<button onClick={() => embla.scrollNext()}>Next</button>
```

### EmblaRef vs Render props
The current implementation with **EmblaRef** allows for more freedom. If any of you think **render props** should be supported side by side I'd appreciate that you open a new issue and a related pull request. 

### useRef & useState
Let's vote below 👇. Give the suggestion you prefer a thumbs up --> 👍.

**Comment by davidcetinkaya** [comment]
Implemented in version [1.2.1](https://github.com/davidcetinkaya/embla-carousel-react/releases/tag/v1.2.1). Thank you all for the discussion that led to these improvements 🎉.

**Comment by davidcetinkaya** [comment]
> I'm still not 100% satisfied by the API, the ref handling still looks a bit weird to me. Somehow the API should provide directly an embla object (using the render props pattern, or providing a useEmbla hook?) to make things even simpler... maybe?

@michaelrambeau check out release [v1.3.0](https://github.com/davidcetinkaya/embla-carousel-react/releases/tag/v1.3.0). The `useEmblaCarousel` hook has now been released thanks to @gunnarx2  🙂.
